bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
-        let tape_size = tape_size.map(NonZeroUsize::get).unwrap_or(30_000);
+    ///
+    /// ```no_run
+    ///# fn main() -> Result<(), Box<dyn std::error::Error>>{
+    ///# use bft_types::BfProgram;
+    ///# use bft_interp::VirtualMachine;
+    ///# use std::num::NonZeroUsize;
+    ///#
+    /// let bf_program = BfProgram::from_file("my_bf_program.bf")?;
+    ///
+    /// let tape_size: Option::<NonZeroUsize> = Some(NonZeroUsize::new(30000).unwrap());
+    /// let bf_interpreter: VirtualMachine<u8> = VirtualMachine::new(&bf_program, tape_size, true);
+    ///#
+    ///# Ok(())
+    ///# }
+    /// ```
+    pub fn new(
+        program: &'a BfProgram,
+        tape_size: Option<NonZeroUsize>,
+        tape_can_grow: bool,
+    ) -> Self {
+        let tape_size = tape_size.map(NonZeroUsize::get).unwrap_or(30_000);
 
         Self {
-            cells: vec![T::default(); tape_size],
-            head: 0,
+            cells: vec![T::default(); tape_size],
+            head: 0,
             tape_can_grow,
+            program,
+            program_counter: 0,
+        }
+    }
+
+    /// Interpret the program
+    ///
+    /// ```no_run
+    ///# fn main() -> Result<(), Box<dyn std::error::Error>>{
+    ///# use bft_types::BfProgram;
+    ///# use bft_interp::VirtualMachine;
+    ///# use std::num::NonZeroUsize;
+    ///#
+    /// let bf_program = BfProgram::from_file("my_bf_program.bf")?;
+    ///
+    /// let tape_size: Option::<NonZeroUsize> = Some(NonZeroUsize::new(30000).unwrap());
+    /// let mut bf_interpreter: VirtualMachine<u8> = VirtualMachine::new(&bf_program, tape_size, true);
+    /// bf_interpreter.interpret_program()?;
+    ///#
+    ///# Ok(())
+    ///# }
+    /// ```   
+    pub fn interpret_program(&mut self) -> Result<(), VMError> {
+        while self.program_counter != self.program.instructions().len() {
+            match self.program.instructions()[self.program_counter].instruction() {
+                Instruction::MoveLeft => self.move_head_left()?,
+                Instruction::MoveRight => self.move_head_right()?,
+                Instruction::Increment => self.increment_cell(),
+                Instruction::Decrement => self.decrement_cell(),
+                Instruction::Input => self.read_value(&mut std::io::stdin())?,
+                Instruction::Output => self.print_value(&mut std::io::stdout())?,
+                _ => (),
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Move the head one cell towards the left (start) of the tape
+    fn move_head_left(&mut self) -> Result<(), VMError> {
+        match self.head.checked_sub(1) {
+            Some(new_pos) => {
+                self.head = new_pos;
+                Ok(())
+            }
+            None => {
+                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                Err(VMError::HeadUnderrun(bad_instruction))
+            }
         }
     }
```

Comment:
```suggestion
    fn move_head_left(&mut self) -> Result<(), VMError> {
        if self.head > 0 {
            self.head -= 1;
            Ok(())
        } else {
            Err(VMError::HeadUnderrun(self.program.instructions()[self.program_counter].clone()))
        }
    }
```

(Above would need checking and also formatting, but I feel it's slightly easier for a human to grok than the use of `checked_sub()`)